File: toolbox/musiclib/lyrics.py
import logging
import re

import requests
from mutagen.id3 import ID3, SYLT, Encoding
from mutagen.mp3 import MP3

log = logging.getLogger("musiclib.lyrics")

LRCLIB_API = "https://lrclib.net/api"


def has_lyrics(mp3_path: str) -> bool:
    """Check whether an MP3 already has a SYLT (synced lyrics) frame."""
    try:
        audio = MP3(mp3_path, ID3=ID3)
        if audio.tags is None:
            return False
        return any(frame.startswith("SYLT") for frame in audio.tags)
    except Exception:
        return False


def parse_lrc(lrc_text: str) -> list[tuple[str, int]]:
    """Parse LRC text into ``[(text, timestamp_ms), ...]`` for SYLT."""
    lines: list[tuple[str, int]] = []
    for line in lrc_text.splitlines():
        m = re.match(r"\[(\d+):(\d+(?:\.\d+)?)\](.*)", line)
        if m:
            mins, secs, text = m.groups()
            ms = int(float(mins) * 60000 + float(secs) * 1000)
            lines.append((text.strip(), ms))
    return lines


def fetch_lyrics(
    title: str, artist: str, album: str
) -> list[tuple[str, int]] | None:
    """Fetch synced LRC lyrics from lrclib.net. Returns parsed SYLT list or None."""
    params = {"track_name": title, "artist_name": artist, "album_name": album}
    try:
        resp = requests.get(f"{LRCLIB_API}/search", params=params, timeout=10)
        log.debug("lrclib search %r → HTTP %s", title, resp.status_code)
        if resp.ok:
            for result in resp.json():
                lrc = result.get("syncedLyrics") or ""
                if lrc:
                    log.debug("lrclib found synced lyrics for %r", title)
                    return parse_lrc(lrc)
        log.debug("lrclib: no synced lyrics for %r", title)
    except Exception as e:
        log.warning("lrclib fetch failed for %r: %s", title, e)
    return None


def embed_lyrics(mp3_path: str, sylt_lines: list[tuple[str, int]]) -> None:
    """Write a SYLT frame into an existing MP3 file."""
    audio = MP3(mp3_path, ID3=ID3)
    if audio.tags is None:
        audio.add_tags()
    audio.tags.add(
        SYLT(
            encoding=Encoding.UTF8,
            lang="eng",
            format=2,  # milliseconds
            type=1,  # lyrics
            desc="",
            text=sylt_lines,
        )
    )
    audio.save()

File: backend/src/lyrics.rs
use regex::Regex;
use serde::Deserialize;
use std::sync::LazyLock;

const LRCLIB_API: &str = "https://lrclib.net/api";

static LRC_LINE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\[(\d+):(\d+(?:\.\d+)?)\](.*)").unwrap());

#[derive(Deserialize)]
struct LrclibResult {
    #[serde(rename = "syncedLyrics")]
    synced_lyrics: Option<String>,
}

/// Parse LRC text into (text, timestamp_ms) pairs for SYLT embedding.
fn parse_lrc(lrc: &str) -> Vec<(String, u32)> {
    lrc.lines()
        .filter_map(|line| {
            let caps = LRC_LINE.captures(line)?;
            let mins: f64 = caps[1].parse().ok()?;
            let secs: f64 = caps[2].parse().ok()?;
            let ms = (mins * 60_000.0 + secs * 1_000.0) as u32;
            let text = caps[3].trim().to_string();
            Some((text, ms))
        })
        .collect()
}

/// Fetch synced lyrics from lrclib.net. Returns parsed SYLT list or None.
pub async fn fetch_lyrics(title: &str, artist: &str, album: &str) -> Option<Vec<(String, u32)>> {
    let client = reqwest::Client::new();
    let resp = client
        .get(format!("{LRCLIB_API}/search"))
        .query(&[
            ("track_name", title),
            ("artist_name", artist),
            ("album_name", album),
        ])
        .timeout(std::time::Duration::from_secs(10))
        .send()
        .await
        .ok()?;

    if !resp.status().is_success() {
        tracing::debug!("lrclib search {title:?} → HTTP {}", resp.status());
        return None;
    }

    let results: Vec<LrclibResult> = resp.json().await.ok()?;
    for result in &results {
        if let Some(lrc) = &result.synced_lyrics {
            if !lrc.is_empty() {
                tracing::debug!("lrclib found synced lyrics for {title:?}");
                return Some(parse_lrc(lrc));
            }
        }
    }

    tracing::debug!("lrclib: no synced lyrics for {title:?}");
    None
}
